### PR TITLE
Added error handling for io_uring creation failure

### PR DESCRIPTION
**Pull request by 1Jo1**
Motivation:

we should throw a  java runtime exception for io_uring creation failure to avoid a `NullPointerException`

Modifications:

-Error handling for creation ring fd and mmap io_uring ring buffer
-Cleanups

Result:

better error handling

### CODE REVIEW

File: transport-native-io_uring/src/main/c/netty_io_uring_native.c [hunk 1]
```diff
@@ -276,28 +244,22 @@ static void netty_io_uring_ring_buffer_exit(JNIEnv *env, jclass class, jobject r
...
     struct io_uring io_uring_ring;
-    //Todo memset instead
-    io_uring_ring.flags = 0;
-    io_uring_ring.sq.sqe_tail = 0;
-    io_uring_ring.sq.sqe_head = 0;
-    setup_io_uring(ring_fd, &io_uring_ring, &p);
+    int ret = setup_io_uring(ring_fd, &io_uring_ring, &p);
+
+    if (ret != 0) {
+        netty_unix_errors_throwRuntimeExceptionErrorNo(env, "failed to mmap io_uring ring buffer", ret);
```
Comment (normanmaurer) [inline transport-native-io_uring/src/main/c/netty_io_uring_native.c:261]: shouldn't that better include the `errno` ?

File: transport-native-io_uring/src/main/c/netty_io_uring_native.c [hunk 2]
```diff
@@ -276,28 +244,22 @@ static void netty_io_uring_ring_buffer_exit(JNIEnv *env, jclass class, jobject r
...
     struct io_uring_params p;
     memset(&p, 0, sizeof(p));
 
     int ring_fd = sys_io_uring_setup((int)entries, &p);
 
-    //Todo
     if (ring_fd < 0) {
-      printf("RingFd error: %d\n", ring_fd);
-      //throw Exception
+     netty_unix_errors_throwRuntimeExceptionErrorNo(env, "failed to create io_uring ring fd ", errno);
```
Comment (normanmaurer) [inline transport-native-io_uring/src/main/c/netty_io_uring_native.c:254]: nit: "missing space"

File: transport-native-io_uring/src/main/c/netty_io_uring.h [hunk 3]
```diff
@@ -38,9 +38,6 @@ struct io_uring_sq {
     unsigned *array;
     struct io_uring_sqe *sqes;
 
-    unsigned sqe_head;
-    unsigned sqe_tail;
```
Comment (normanmaurer) [inline transport-native-io_uring/src/main/c/netty_io_uring.h:42]: why this change ?

**Review by normanmaurer — CHANGES_REQUESTED** [review]

File: transport-native-io_uring/src/main/c/netty_io_uring_native.c (hunk 1 above)
Comment (1Jo1) [reply transport-native-io_uring/src/main/c/netty_io_uring_native.c:261]: yeah sure, however we already return a errno in `setup_io_uring`, I thought it would be more readable

Comment (normanmaurer) [reply transport-native-io_uring/src/main/c/netty_io_uring_native.c:261]: ah ok .... missed that. Then its fine :)

File: transport-native-io_uring/src/main/c/netty_io_uring.h (hunk 3 above)
Comment (1Jo1) [reply transport-native-io_uring/src/main/c/netty_io_uring.h:42]: we don't use and we only need io_uring_sq struct(which is on the stack) for storing kernel pointers for the RingBuffer creation

**Comment by normanmaurer** [comment]
@netty-bot test this please

**Comment by normanmaurer** [comment]
@netty-bot test this please

File: transport-native-io_uring/src/main/c/netty_io_uring_native.c [hunk 4]
```diff
@@ -276,28 +244,22 @@ static void netty_io_uring_ring_buffer_exit(JNIEnv *env, jclass class, jobject r
...
     struct io_uring io_uring_ring;
-    //Todo memset instead
-    io_uring_ring.flags = 0;
-    io_uring_ring.sq.sqe_tail = 0;
-    io_uring_ring.sq.sqe_head = 0;
-    setup_io_uring(ring_fd, &io_uring_ring, &p);
+    int ret = setup_io_uring(ring_fd, &io_uring_ring, &p);
+
+    if (ret != 0) {
+        netty_unix_errors_throwRuntimeExceptionErrorNo(env, "failed to mmap io_uring ring buffer", ret);
```
Comment (normanmaurer) [inline transport-native-io_uring/src/main/c/netty_io_uring_native.c:261]: you also need to add a return statement as the exception will be just put on the "stack" and picked up once the method returns. This works a bit different then throwing an exception from java layer :)

```
if (ret != 0) {
    netty_unix_errors_throwRuntimeExceptionErrorNo(env, "failed to mmap io_uring ring buffer", ret);
    return NULL;
}
```

**Review by normanmaurer — CHANGES_REQUESTED** [review]

**Review by chrisvest — APPROVED** [review]
Looks fine besides the return that Norman mentioned. Pre-approving assuming that gets fixed.

**Comment by normanmaurer** [comment]
@netty-bot test this please

**Review by normanmaurer — APPROVED** [review]

**Review by njhill — COMMENTED** [review]
:+1: @1Jo1 nice!
